### PR TITLE
fix(lsp): fix bug when deleting LspNotify autocmd for inlay hints

### LINKED ISSUE #24456: inlay_hint: "Failed to delete autocmd" error when closing a buffer
### Problem

With the latest nightly release, I'm seeing the following error every time I close a buffer:

image

So far I've seen this happen with Lua and Rust files.

Note that this doesn't happen with NVIM v0.10.0-dev-692+g86ce3878d.

### Steps to reproduce

The only inlay-hint related setup I have is this pair of autocommands that toggles inlay hints:
```lua
    if buf_client.supports_method 'textDocument/inlayHint' then
        local inlay_hints_group = vim.api.nvim_create_augroup('ToggleInlayHints', { clear = false })

        -- Initial inlay hint display.
        local mode = vim.api.nvim_get_mode().mode
        vim.lsp.inlay_hint(bufnr, mode == 'n' or mode == 'v')

        vim.api.nvim_create_autocmd('InsertEnter', {
            group = inlay_hints_group,
            buffer = bufnr,
            callback = function()
                vim.lsp.inlay_hint(bufnr, false)
            end,
        })
        vim.api.nvim_create_autocmd('InsertLeave', {
            group = inlay_hints_group,
            buffer = bufnr,
            callback = function()
                vim.lsp.inlay_hint(bufnr, true)
            end,
        })
    end
```
The rest of my LSP configuration is [here](https://github.com/MariaSolOs/dotfiles/blob/d0c27368177b1ca78d79843824e9d54f565e94ba/.config/nvim/lua/helpers/lsp.lua).

### Expected behavior

For there to be no error.

### Neovim version (nvim -v)

NVIM v0.10.0-dev-709+gce56ad2ba

### Vim (not Nvim) behaves the same?

No

### Operating system/version

Mac Ventura 13.4.1

### Terminal name/version

kitty 0.28.1

### $TERM environment variable

xterm-kitty

### Installation

bob

### PR DESCRIPTION
**Pull request by catlee**
This PR changes a few things in the inlay_hint and diagnostic LSP code.

1. It re-introduces the `enabled` flag for the buffer state tables. Previously I was relying on the presence of an autocmd id in the state table to track whether inlay_hint / diagnostic was enabled for a buffer. There are two reasons why this doesn't work well:
  a. Each time inlay_hint / diagnostic is enabled, we call `nvim_buf_attach` on the buffer, resulting in multiple `on_reload` or `on_detach` callbacks being registered.
  b. Commands like `bwipeout` delete buffer local autocmds, sometimes before our `on_detach` callbacks have a chance to delete them first. This causes the error referenced in #24456.
2. Change the `LspDetach` autocmd to run each time (i.e., I removed the `once` flag from it). Since we're only registering autocmds once per buffer now, we need to make sure that we set the enabled flag properly each time the LSP client detaches from the buffer.
3. Have the `LspNotify` handler also refresh the inlay_hint / diagnostics when receiving the `textDocument/didOpen` event. Before this point, the LSP backend doesn't have the contents of the buffer, and so can't provide inlay hints or diagnostics.

Downsides of this approach:
* When inlay_hint / diagnostics are disabled on a buffer, it will continue to receive `LspNotify` events for that buffer. The callback exits early since the `enabled` flag is false.

Alternatives:
* Can we wrap the call to `nvim_del_autocmd` in `pcall` to swallow any errors resulting from trying to delete the autocmd?

Fixes #24456

### CODE REVIEW

**Comment by catlee** [comment]
I'll need to do the same for `diagnostics.lua` - you can hit the same problem with `:bwipeout!` causing the plugin to fail to delete the autocmd

File: runtime/lua/vim/lsp/inlay_hint.lua [hunk 1]
```diff
@@ -136,24 +138,27 @@ local function enable(bufnr)
...
-    bufstates[bufnr] = { applied = {} }
-    bufstates[bufnr].autocmd_id = api.nvim_create_autocmd('LspNotify', {
+    bufstates[bufnr] = { applied = {}, enabled = true }
+    api.nvim_create_autocmd('LspNotify', {
       buffer = bufnr,
       callback = function(opts)
-        if opts.data.method ~= 'textDocument/didChange' then
+        if
+          opts.data.method ~= 'textDocument/didChange'
+          and opts.data.method ~= 'textDocumen/didOpen'
```
Comment (MariaSolOs) [inline runtime/lua/vim/lsp/inlay_hint.lua:147]: Typo:
```suggestion
          and opts.data.method ~= 'textDocument/didOpen'
```

File: runtime/lua/vim/lsp/diagnostic.lua [hunk 2]
```diff
@@ -416,38 +415,46 @@ function M._enable(bufnr)
...
-    return
+  if not bufstates[bufnr] then
+    bufstates[bufnr] = { enabled = true }
+
+    api.nvim_create_autocmd('LspNotify', {
+      buffer = bufnr,
+      callback = function(opts)
+        if
+          opts.data.method ~= 'textDocument/didChange'
+          and opts.data.method ~= 'textDocument/didOpen'
```
Comment (MariaSolOs) [inline runtime/lua/vim/lsp/diagnostic.lua:426]: It would be nice to be able to use the LSP method names as patterns for the autocommand...

**Comment by glepnir** [comment]
could write test case for using `LspNotify` on `inlayhint`?

File: runtime/lua/vim/lsp/diagnostic.lua (hunk 2 above)
Comment (catlee) [reply runtime/lua/vim/lsp/diagnostic.lua:426]: Yeah, unfortunately patterns can't be used with buffer local autocmds:
https://neovim.io/doc/user/api.html#api-autocmd

**Comment by stasjok** [comment]
Didn't test it, but looks like if there are two language servers supporting pull diagnostics, stopping one of them will disable it completely (after LspDetach event). Probably it should disable itself only after all servers supporting given method detached (or count a number of attached servers).

**Comment by catlee** [comment]
> Didn't test it, but looks like if there are two language servers supporting pull diagnostics, stopping one of them will disable it completely (after LspDetach event). Probably it should disable itself only after all servers supporting given method detached (or count a number of attached servers).

Hmm, that's true.
Perhaps we need to store state per client?

**Comment by MariaSolOs** [comment]
Hello! Is there a specific reason why this hasn't been merged yet...?
I'm happy to help in any way I can btw :)

**Review by glepnir — APPROVED** [review]

**Comment by glepnir** [comment]
cc @mfussenegger

**Pull request merged**